Support for helpline sessions

liewegas opened the pull request:
The key idea here is to separate different conversations with the helpline into different "sessions".  Sessions would most commonly be separated by elections, but they could also (in the future) be triggered if a long time has simply passed since the voter last interacted with the helpline.

When this code is deployed, all existing sessions will lack the sessionStartEpoch field, and will be considered "stale".  A stale session is a session that is in a sort of limbo state: we have decided that the sessions is over, but all of the blocks in slack don't reflect that fact.  This is because we'll have thousands and thousands of conversations leading up to an election, and will then do a mass-update in the database to, e.g., reset the voter status to Registered from Voted, but we don't currently iterate over all of the slack state to make it reflect that.

But, fear not--stale sessions show themselves as soon as you interact with them.

If a volunteer tries to text a voter via a stale session thread, they'll get an error message, and a helpful prompt to use one of the new commands (!resume-session or !new-session).  If a voter texts the helpline, they'll get the generic welcome text (as if they are a new user), and at that point the slack state will get refreshed so that the blocks says "This session has ended".

If a voter reenters the helpline, past sessions are visible in the history.  The existing history block is still there showing *just* messages for the new session (and whatever preceded HELPLINE).  Above that there are one or more messages that indicate there was a previous session and how long ago, along with buttons to Show or Hide it.  If the history is too long for a single slack message block, there are buttons to paginate.

If a volunteer encounters a stale thread, they have two choices:

- ``!resume-session`` will resume the existing session.  The blocks will refresh (pulling the latest voter status from the db), and the volunteer assignment will be unchanged.  The volunteer can then text the voter and the voter can respond as if nothing happened.
- ``!new-session channelname`` will open up a new session with the voter, create a new thread in an arbitrary channel (could be the same one!).  The voter state is refreshed and the volunteer assignment is cleared.  The old sessions history is of course visible.  The volunteer can immediately start texting the voter and the voter can respond, and the voter won't realize anything happened (no need to send HELPLINE to reenter), so the conversation can continue.

These two options, from the voter's perspective, look the same.  The only difference is on the volunteer side of things.  Is this a continuation of the earlier conversation, or a fresh conversation about a new upcoming election?  Is it more convenient to (1) clear the volunteer and (2) route the voter to a new channel all in one go?

There are a few other things that were added along the way that are unrelated to sessions:

- ``!route channelname`` now works in any voter thread.  This is way more convenient than the old ``@helpline ROUTER_VOTER cut-and-paste-id 48298 channelname`` flow.  The behavior is the same, including the check mark on success and error message in the thread if there is a problem.
- ``!fake-old-session`` will make a session *look* like a pre-sessions thread (with no sessionStartEpoch) for the purposes of testing.  This just cleared the field from redis.
- A series that adds ``twilio_phone_number`` to the ``threads`` table precedes all of this, which gets us away from relying on ``is_demo`` while the reset of the code uses user_id + twilio_phone_number.
- The voter header is shorter and generated from a single helper function.  (Push vs pull is uninteresting because it never changes within an org, and the userId + twilio number can share the same line.)

There are some other cosmetic improvements along the way (more X and white_check_mark reactions; links to new location of session/thread).

There is also some ick:

- Most of the existing postgres fields are TIMESTAMP and not TIMESTAMPTZ, which makes things gross with respect to timezone.  In order for conversion from postgres timestamp back to unix epoch to not offset by timezone offset each time, the new fields are TIMESTAMPTZ.  Probably all of the existing fields should have been, but they're not. In one case a select casts to ::timestamptz to work around this, but perhaps a future cleanup will convert everything to TIMESTAMPTZ... and also switch to UTC everywhere! But I leave that to future generations.
- Slack has a maximum message size of 3k-4k (not sure exactly), which prevents us from putting too much history content in one message. But our past session history is a single block that we update, so we have pagination.  We are also limited to like 5 buttons across, so it always shows first/last, current, and the next/prev. Even though the button content has a `type`, it only supports `plain_text`, not `mrkdwn`, so the current position looks like ``*2*`` (vs ``1`` and ``3`` for adjacent).  Clicking the current *2* button doesn't do anything, but you can't mix buttons and non-buttons in the blocks.  I don't think it matters.
- Also, if you expand blocks with lots of text, Slack does it's own "Show more" and "Show less" thing which is kind of annoying when you're literally clicking a "Show" button.. but I don't think there is a way around that.  It's only slightly annoying.

TODO:
- [x] fix router tests
- [x] note whether voter is new or returning (add returningVoter boolean to UserInfo)

A new thread with past sessions:

![image](https://user-images.githubusercontent.com/433031/99133505-df58b700-25df-11eb-9739-2760198348b6.png)

Expanded short session:

![image](https://user-images.githubusercontent.com/433031/99133539-f8f9fe80-25df-11eb-91c5-bc1fdfa00ab0.png)

Expanded with pagination:

![image](https://user-images.githubusercontent.com/433031/99133552-0e6f2880-25e0-11eb-8c86-fe466dd8eb2a.png)

Volunteer texting to a stale session:

![image](https://user-images.githubusercontent.com/433031/99133569-26df4300-25e0-11eb-8462-bc132e2de7d5.png)

liewegas committed:
router: checkmark on route success instead of reply

This avoids cluttering up your threads view

liewegas committed:
add twilio_phone_number to threads column

This is primarily to be consistent with the rest of the code base.

liewegas committed:
stop ignoring errors accessing the threads table

These were only needed for the transition period while creating +
populating the threads table.

liewegas committed:
make postUserMessageHistoryToSlack() return a promise with the ts

liewegas committed:
add historyTs to ThreadInfo

liewegas committed:
userInfo: add start time, and start fresh session when encountering old infos

liewegas committed:
make DatabaseThreadEntry fields non-null

liewegas committed:
fix history_ts populate after routing a voter

This fixes a recent regression!

liewegas committed:
set session_end_at when we create a fresh userInfo

liewegas committed:
set session_start_at for new threads

liewegas committed:
getPastSessionThreads() helper

liewegas committed:
improve error message for stale threads: link to new channel AND thread

liewegas committed:
populate initial blocks with any prior status (not just VOTED)

liewegas committed:
ignore slack actions on inactive threads (and react with zombie)

liewegas committed:
close blocks on old session/thread

liewegas committed:
pass redis client, userInfo to postUserMessageHistoryToSlack

liewegas committed:
preserve session start time when we handleNewVoter

The first UserInfo is created when we get the first message, but
handleNewVoter *also* creates a new one.  (This is I think because there
are multiple callers... and one of them doesn't have a UserInfo.)

Pass in our original userInfo (if we have one) so that we don't ignore
the part of the message history before they texted HELPLINE.

Also, fix the start time: (1) correct for TZ (ugh why don't we use UTC
everywhere?) and (2) pad by 10 seconds so that we capture the most recent
incoming message.

liewegas committed:
include session start/end in ThreadInfo

liewegas committed:
show prior sessions separately in the message history

liewegas committed:
fix behavior when we route back to channel with an existing thread

- push logic on picking the start timestamp and preamble/context text
down into the helper, simplifying arguments in the callers
- only post the new messages when returning to a thread--not the older
message or old sessions

liewegas committed:
only relay slack message from the current thread

This additional condition ensures that if a voter was in a channel,
time/election goes by and they get a fresh session, and they end up in
the same channel, sending an outgoing slack message on the old thread will
fail.

liewegas committed:
when routing back to a channel, link to the thread

The old message required you to manually search for the thread.

liewegas committed:
paginate past session history

liewegas committed:
add threads columns for session start/end

liewegas committed:
support '!route channelname' inside the thread

This is SO much easier to use.  Only available to admins.

liewegas committed:
move stale session check into helper isStaleSession()

Later we'll have other conditions here (besides pre-2020 sessions)

liewegas committed:
introduceNewVoterToSlackChannel: optional outgoing message; return ts

liewegas committed:
DbApiUtil: ignore threads for sessions that have ended

liewegas committed:
reset volunteer when ending an old session

liewegas committed:
migration/query to reset voter status after each election

liewegas committed:
x reaction on bad !route command

liewegas committed:
make 'Routing voter to ...' messages link to destination channel

liewegas committed:
add some slop for session end time

Otherwise, if message seconds is 12.3, session end is 12.5, the end
gets rounded down to 12 and the final message is missed in the session
history.

liewegas committed:
label past sessions with absolute (not relative) time

liewegas committed:
leave *Operator* prefix off voter session/message history messages

It's clear from context (and the icon), and it makes it harder to fit
the top line on a single line.

liewegas committed:
include <!channel> when known number voter is introduced

liewegas committed:
eslint cleanups

liewegas committed:
construct userInfo in handleNewVoter and welcomePotentialVoter caller

This puts it all in app.ts, where we have more context.

This lets us drop the sessionStartEpoch workaround since we aren't creating
a new userInfo in that path.

liewegas committed:
note whether voter is new or returning

Also consistently bold the state

liewegas committed:
standardize voter header and generate it from a single helper

Shorten to 1 line, removing extraneous info (push vs pull).

liewegas committed:
fix resetDemo

- reset all threads and sessions
- find ts from message interactions

liewegas committed:
react when we get a slack message on an archived thread

liewegas committed:
route a voter => volunteerEngaged

Otherwise, a fresh voter could get routed by a volunteer to a state
channel, and if the next message is from the voter it will get
interpreted as a state attempt.

Don't put this in routeVoterToSlackChannel() since one caller is
automated and setting this value doesn't make sense in that case.

liewegas committed:
constently use 'archived IS NOT TRUE'

For consistency and correctness!

- For the messages queries, this mostly doesn't matter, because we set
archived=false.
- For getSlackThreadsForVoterAllSessions, this matters, because archived
is null for non-archived threads
- For getThreadLatestMessageTs, this doesn't matter, because we are looking
for the newest timestamp only.

liewegas committed:
setSessionEnd: only end sessions that aren't archived

(This is mostly data hygiene... there is not effect on behavior.

liewegas committed:
formatting

liewegas committed:
getThreadLatestMessageTs doesn't depend on thread archived/active state

This is querying a specific channel,thread combo.

liewegas committed:
consistently check active, session_end_at, archived to queries

Put these 3 checks (when appropriate) at the top of each query so they
are always in the same place and in the same order.

Subtley fixes most of these; cosmetic cleanup to the others.

liewegas committed:
no history message cleanup

liewegas committed:
!fake-old-session, !new-session, !resume-session commands in thread

Only for admins!

benweissmann commented on src/db_api_util.ts:
```diff
@@ -581,25 +606,71 @@ export async function archiveDemoVoter(
+  try {
+    await client.query(
+      `UPDATE threads SET session_end_at = (
+        SELECT MAX(updated_at) + interval '1 second' FROM threads
```
What's the 1 second for?

liewegas replied:
The message might be 11.2 seconds, the updated_at 11.3 seconds, and when we do the select it ends up rounded down to 11, which means we leave off the the last message in the session.  The 1 second just provides a buffer.

benweissmann replied:
got it -- can you just add a comment to that effect?

benweissmann commented on src/router.ts:
```diff
@@ -31,20 +31,57 @@ type UserOptions = {
 
+export function isStaleSession(userInfo: UserInfo): boolean {
+  if (!userInfo.sessionStartEpoch) {
+    // legacy session from before 2020-11-03 election
```
This isn't specific to 2020-11-03, right? It's just any session that's been ended by running that SQL migration?

liewegas replied:
It's specific to anything that predates this branch being merged, since starting now we'll populate the sessionStartEpoch when creating a new UserInfo.

benweissmann commented on src/router.ts:
```diff
@@ -1372,7 +1541,237 @@ export async function handleSlackVoterThreadMessage(
   )) as UserInfo;
   // Only relay Slack messages from the active Slack thread.
-  if (userInfo.activeChannelId === reqBody.event.channel) {
+  if (
```
Can we pull this stuff out into a separate function or file? This function is getting pretty hard to follow -- maybe something like `if (handleCommand(messageToSend)) { return; }` and then `handleCommand()` can return `true` if the message was a command and `false` if it was a normal message.

benweissmann replied:
And then that function can help DRY out some of this, like handling the ACL check at the top of each of these branches

liewegas replied:
Yep--updated!

benweissmann reviewed (changes requested):
Looks good, just a few suggestions / questions

liewegas committed:
move in-thread commands into a helper

Simplify the ! command logic a bit: anything that starts with ! is a
command, unless it is all !'s (e.g., '!!!' is okay to send to a voter).

This consolidates the admin checks into one place

benweissmann reviewed (approved):
looks great!
